373737 konfigurierbare Kachel-Tabellen-Anker

mgollwitzer-bsi opened the pull request.

jmiserez committed:
DynamicResourceInfo: Allow internal loading of BinaryResource from URL

Add method to DynamicResourceInfo to support the loading of a
BinaryResource only by IUiSession, without knowing the IJsonAdapter
directly.

329612

(cherry picked from commit 89ec87843c6364c76cb473fd45ddfcd63631258f)

egloff commented on eclipse-scout-core/src/table/Table.less:
```diff
@@ -655,3 +655,19 @@
   padding: 0;
 }
+
+.anchored-table {
```
This css-class is only used in our internal crm-repo and therefore should be defined there...

egloff commented on eclipse-scout-core/src/table/TableTileGridMediator.ts:
```diff
@@ -294,7 +294,7 @@ export class TableTileGridMediator extends Widget implements TableTileGridMediat
     return scout.create(TileAccordion, {
       parent: this.table,
-      virtual: true,
+      virtual: false,
```
I don't think it is good to change the default behavior of all TileAccordions created by the TableTileGridMediator to "not-virtual". This should be set via a property only where it is needed (analogue gridColumnCount).

aschwanden committed:
390626 BeanManager: Concurrency Issue when unregistering

bsifpa committed:
Remove disabled style for SimpleTab

cguglielmo committed:
Resizable: wrong cursor during resizing

The resize cursor should be visible during the whole resize process.
Currently, it is only visible when the cursor is on the resizable
handle.

Also simplified Resizable.ts a little bit and made sure resize overlay
is removed when resizable is destroyed.

389402

cguglielmo committed:
TileResizeHandler: make resizing more reactive

The resizing now reacts on the gaps between the cells as well.

389402

cguglielmo committed:
Resizable/MoveSupport: cancel resizing when pressing ESC

389402

cguglielmo committed:
Resizable: don't use the same event name as the browser

There is a native 'resize' event triggered by the browser when the
window is resized.
Because the resizable element triggers an event with the same name on a
DOM element, it bubbles up to the window executing the event handlers
listening for window resizes, which is unexpected.
The event handlers could check the target, but since this is an
artificial event and not the same as the one sent by the browser, the
handlers should not need to take care of that.

dwl-bsi committed:
Extensions: use extension chain in AbstractSmartColumnExtension

The method execPrepareLookup in the AbstractSmartColumnExtension must
use the SmartColumnPrepareLookupChain to call execPrepareLookup of the
chain. Otherwise, it is not possible to call the original
execPrepareLookup method of the extended SmartColum.

391224

fschinkel committed:
Move MetaResource to scout.rt

Move MetaResource, MetaVersionInfoDo and IApiVersion to scout.rt.
Implementations of IApiVersion need to be moved to rest.server module.

cguglielmo committed:
Resizable/MoveSupport: only allow resizing with the left mouse button

The right mouse button typically opens the context menu.
If resizing can be started with that button, the context menu would open
as well, which is confusing.

Also ensure the resize overlay won't stay in the DOM, e.g. if the right
button is pressed while resizing has already started, or mouse is
released outside the window without triggering a mouseup event.

389402

egloff committed:
Bump BSI Customer Suite version to 25.1

cguglielmo committed:
FormFieldTile: remove disabled style

The current disabled style looks bad, especially the inverted one.
Since most of the tiles (except button like tiles) can be used when
disabled anyway, they should not look disabled.

-> Removed disabled style of regular form field tiles and adjusted
   disabled style of tile buttons.

Also set tile selection border width of inverted tiles in dark mode
to 2px because inverted and regular tiles look the same.

389352

mvilliger committed:
Skip remote files certificates in UI server

The trust manager is started during platform startup. At this time
the remote files cannot be fetched from the backend because:
1. the backend might not yet be started
2. a RunContext is missing which is required to use the ServiceTunnel

Therefore, skip remote files in the UI when setting up the TrustManager.

(cherry picked from commit 4e87645741b88ca42c754bf8dd73bbd332732dd5)

mvilliger committed:
Remove non-existing texts from UiTextContributor

(cherry picked from commit b6ba1bbe7b942909686c5bdb0c1d3a5fa3523c6b)

mvilliger committed:
CodeType Browser Cache: Support user dependent CodeTypes

Only update the browser CodeTypes cache with the notification message
for the user that invalidates the server-side cache. The notification
directly holds the CodeType data so that dependent JS code gets the new
data timely.

All other users only get a message with the CodeType ids to update and
create a new (potentially delayed) request to reload these CodeTypes.

390963

(cherry picked from commit 4fbeb008f24160385d6d2d91f32d38c0fd164df7)

andibur committed:
Testing: improve check of async jobs started in test cases

Add execution hint for background jobs executing deferred operations
that are not directly attached to a particular test case.

391676

(cherry picked from commit 7849976083da572f55d9d3840f17dd92ace24715)

mvilliger committed:
Ensure creation time is unique per node

If multiple notifications are added within the same millisecond the
creation time should still be unique for all notifications.

mfi123 committed:
Improve error text for cyclic dependencies

The dependencies displayed in the exception are only a subset of all
dependencies. Error text improved to avoid confusion.

383291

cguglielmo committed:
CodeTypeCache: make it more robust

- Abort an existing call
  This is necessary to make sure a second request cannot overtake the
  first one.
- Add retry handling
  This is necessary to ensure the cache will eventually be updated even
  if the reload request fails due to network errors.

390963

cguglielmo committed:
FocusManager: ignore elements covered by a glasspane while tabbing

If an element is covered by a glasspane, it must not be tabbable.
Currently, the focus manager correctly prohibits focus changes to
such elements, but it will focus the entry point instead. This means,
if the user is tabbing, the entry point will be focused instead of the
next tabbable element in a focus context.

The change in Widget.isFocusable was not necessary to fix this issue
but it makes it consistent.

bschwarzent committed:
TileGridLayout: mark tiles as 'animating' while changing bounds

Add a promise to the animated container while tile bounds are changed
with a jQuery animation. This allows the HtmlComponent to wait for the
animation to finish before validating the content (otherwise, wrong
dimensions might get measured). It works similarly to the animate-*
classes that are used for CSS animations, which we cannot use in this
case, because no animationEnd would is fired after the jQuery animation
has been completed.

To clean up immediately after the animation, we use the always()
callback. Unlike then(), it is also executed when the animation is
aborted. The callback is run synchronously when stop() is called.

Instead of starting four individual animations and combining them into a
new promise, we use the cssAnimated() function to animate all properties
at the same time. Besides using less code, this prevents an endless loop
which can occur when the animation promises are created with the option
"queue: false" and a "then()" callback is attached. This is most likely
a bug in JQuery.

390356

cguglielmo committed:
Create property change events for message boxes, forms, file choosers

This makes it possible to wait for display children to be opened
which is especially handy when writing tests.

mfi123 committed:
ClientNotificationNodeQueue maximum queue size

Change the default value for the queue size and number of notifications
that are processed at once. Several projects are already using this
larger value, the old value is no longer up to date. In addition, change
the log level from Warn to Error so that no project misses this error.

368114

fschinkel committed:
IdCodec: use IdCodecException

Introduce new IdCodecException and use it in IdCodec whenever an
exception is thrown during serialization/deserialization of an IId.
Also add an IdCodecExceptionMapper, that returns a "400 - Bad Request"
when an IdCodecException is thrown and only logs the stacktrace if
debug is enabled.

391961

fschinkel committed:
SimpleTab: fix rendering the title/subTitle if it is set to null

The title and the subTitle can be set to null, undefined or empty
string. An earlier implementation has handled this case by removing the
DOM elements $title or $subTitle. When this was refactored the removal
of the DOM elements was removed and since then the content of the DOM
element is only updated if the value is not empty. Because of this the
content could not have been removed.

bschwarzent committed:
Override default user agent style for :focus and :focus-visible

Since Chrome 127, an ugly black outline is sometimes visible on some
widgets, e.g. Table, Tree, TileGrid. This seems to be caused by a
default user agent style for :focus-visible which is somehow applied
more frequently. Because we use our own custom focus style for
focusable widgets anyway, we might as well hide the default outline
border globally.

381281

gabrielstalder committed:
Clean up code and javadoc for decimal columns

bschwarzent committed:
Form: extract computation of cacheBounds storage key

Allow easier customization.

bschwarzent committed:
Remove obsolete prefixes

bschwarzent committed:
Form: improve position of key-box for 'closer' icon

If a form has no close button, position the key-box (F1) over the
closer icon. Otherwise it will be cut off, because the form header
has 'overflow: hidden'.

andibur committed:
HttpProxy: close the target's input stream in case of an exception

The HttpProxy uses Apache HTTP Client and its connection pool. It seems
that responses must be fully read in order to put the leased HTTP
connection back to the pool. Otherwise, the pool starves and the proxy
stops working.

(cherry picked from commit 487dffe74adec5a7a7e138f4882573344e1158f4)

bschwarzent committed:
ComboMenu: fix prefSize when child actions have a custom layout

A ComboMenu may contain a FormFieldMenu which in turn contains a
FormField, e.g. a Button. When the MenuBar is layouted initially, it
reads the preferred size of each menu item before they are actually
layouted. The NullLayout just reads the current CSS size from the DOM.
This is not necessarily correct if the inner layout (e.g. a Button)
relies on its layout to report the correct preferred size. To fix this,
we explicitly set a ColumnLayout to the ComboMenu that has a different
implementation of the preferredLayoutSize() method.

389822

andibur committed:
UiNotificationRegistry: fix potential deadlock when notifying listeners

putInternal notified listeners while holding the write lock (i.e. an
open call). With the current primary use case, this led to deadlocks:
The async servlet request was responded, but the thread did not return
until the response was completely processed by the client. In some cases
the client did not consume the message. Hence, the listener did not
complete and kept the writeLock. Restarting the UI server did close the
network connection. As a result, the listener completed (with an
exception) and finally released the write lock (the client-side issue
is already addressed by 487dffe74adec5a7a7e138f4882573344e1158f4).

Further, the commit slightly improves the performance of cleanup.

393563

andibur committed:
CodeResource: switch locale using NlsLocale instead of a new RunContext

Depending on the application's RunContextFactory, copying the current
RunContext can create a new transaction as well and transaction-bound
resources must be obtained again. Typically, these resources are also
limited and managed by a pool (e.g. DB connections). If the outer and
inner transactions require a limited resource, deadlocks can occur.

393747

mvilliger committed:
Add Jackson to server.app

Now always required in UiNotificationClusterMessage as soon as the first
UiNotification is added to the registry.

mgollwitzer-bsi committed:
Fix toValidFilename for malformed filenames

Filenames in the form of "...ext" were transformed to "ext", leading
problems when the extension was relevant for the further use.

Solution is to keep the extension intact and only replace all but the
dot right before the extension with '_'.

357297

cguglielmo committed:
Menubar: don't wrap right menu box

It can happen that the right menu box wraps if it is slightly too large
and does not fully fit into the menubar.

The bug was introduced with 24.1 with the change of the DOM order of the
menu boxes due to accessibility reasons.
It looks like wrapping cannot be controlled easily when using
float: right, so it now uses display: flex and margin-left: auto which
seems to work reliably.

388293

nsteger committed:
Do not use filterMenuRec for Calendar menus

MenuUtility.filterMenusRec(List<IMenu>, Predicate<IMenu>) wrappes menus
with child actions. In the case of the calendar, no wrapper menus should
be used, since wrapper menus won't propagate the owner value change
event.
Added new method MenuUtility.filterMenus(List<IMenu>, Predicate<IMenu>)
which is used by the calendar instead

332126

andibur committed:
HttpProxy: do not use ExceptionHandler in case of failed HTTP requests

393617

(cherry picked from commit 09cb7b3645752c3777b724c0ad4998f02bd834e1)

mvilliger committed:
Permission UiNotification: only notify affected users

If the server-side permissions cache is invalidated using the
KeyCacheEntryFilter (which only invalidates specific entries), only
users that belong to these invalidated entries are notified.

This is important as the cache entry for a user is invalidated on
logout which would refresh all users browser caches when a user closes
the application.

393577

Knicksaas committed:
Add shake animation when unchecking last visible calendar

When the user tries to hide the last visible calendar, the tree node
shakes and is not uncheckable. This commit includes also a bugfix to
correctly update the visibility of a calendar when the calendar is
initialized but not rendered.

387614

cguglielmo committed:
TileButton: support labelVisible

Tile buttons did not respect the label visible property.
The same applies to regular buttons, but since they are used frequently
changing their behaviour would likely break existing code (see also
TODO in Button.ts).

393740

matthiaso committed:
Increase version for jakarta-activation/mail

Also:

* Remove duplicate dependencies
* Add exclusions for former used com.sun.[activation, mail] dependencies
* Remove exclusions for newer dependencies
* Include Eclipse Angus implementation

391137

paolobazzi committed:
HTTP client connection manager pools metrics

Add OpenTelemetry metrics for HTTP client connection pool:
- active/idle connections
- max connections

Co-authored-by: Bruno Koeferli <bruno.koeferli@bsiag.com>

393561

(cherry picked from commit ab97e46f21e4f3f4900f0318ae0d0182ddba213f)

ab97e46f 393561 HTTP client connection manager pools metrics

Co-authored-by: Paolo Bazzi <paolo.bazzi@bsi-software.com>

paolobazzi committed:
Connection properties for REST HTTP client

Added config properties and REST client properties for:
- max connections in total
- max connections per route
- connection keep alive time
- connection validate after inactivity time

393566

Knicksaas committed:
Calendar: don't apply selection twice when clicking on component

When a click happened on a calendar component, two event listeners have
changed the calendar selection. Only one selection change is necessary.
Both listeners handle now the MouseDown Event and the calendar is
ignoring the click when it happens on a component, only the component
is changing the selection.

394064

ralph-steiner committed:
Log cache invalidations

andibur committed:
ServerSessionProviderWithCache: stop sessions on eviction

393979

(cherry picked from commit 8af7e5326391e6ef316ec5deefc555994d6d1f95)

matthiaso committed:
HttpProxy: Avoid causing another exception during exception handling

Apparently throwing another exception while handling a previous
exception (within failed) method may corrupt some connection pools
[IOReactorShutdownException for following requests].

394532

cguglielmo committed:
TileButton: support keyStroke and preventDoubleClick properties

387468

cguglielmo committed:
Button: update submenu icon dynamically

Submenu icon depends on menus, icon and label -> update the submenu icon
whenever one of the other properties change

Also:
- removed duplicated aria code, it is already done by _updateMenus
- removed duplicate menu keystroke registration, it is already done
  by the form field, see _setMenus

rogersteblerbsi committed:
JAX-WS: Add authentication handlers to bean manager

279344

mvilliger committed:
Reduce number of CodeType to DataObject conversions

Only convert CodeTypes to DataObjects if required:
- when requesting CodeTypes by Id using the CodeResource
- when creating UI notifications for CodeTypes invalidated in the cache

394632

rogersteblerbsi committed:
MailHelper: extraction of Message-ID header should be case-insensitive

350374

rogersteblerbsi committed:
HealthCheck: Log detailed warning in case health check is not successful

361571

mvilliger committed:
UI notification should only be sent once in a cluster

In a cluster environment the UI notification for an event should only be
sent to the browser once. Therefore, do not send UI notifications to
other cluster nodes if the underlying trigger is already cluster-aware.

394718

andibur committed:
Permission cache: invalidate on logout only if triggered by HTTP request

The HTTP session timeout of the backend server is small compared to
the lifetime of the HTTP session on the UI server (default 5 minutes).
Whenever a backend session is invalidated due to inactivity, the
permissions of the affected user are also removed from the
AccessControlCache. This leads to a client notification and now also to
a UI notification. The UI server removes the user from its cache and the
web browser reloads the permissions. Both although the user is still
logged in and his permissions did not change.

It is correct that the AccessControlCache is invalidated in the event of
a dedicated logout by the user.

It is also desirable that a session timeout on the UI server leads to a
complete logout on the backend server, including invalidation of the
AccessControlCache.

394781

(cherry picked from commit 1364422c2461ff3e40cfe1361d08633b2672e52a)

Knicksaas committed:
Calendar: incorrect scroll shadow when on month view

The calendar-scrollable-components class indicates divs, which are
scrollable and are holding components. With the introduction of the
resource calendar, the components are located in a div with class
resource-column. The scrollable class was still on the parent div. This
was fixed in this commit. Also, the full day index stack key calculation
was not working. It should not separate the full day appointments when
the display mode is week or workweek.

394666

cguglielmo committed:
Create base class for display child controllers

The current controllers contain a lot of duplicated code which can be
extracted to a common base class.

matthiaso committed:
Improve logging for HttpProxy async requests

* Copy run monitor (or create a new one) to include correlation id and
further diagnostics information.

374142

matthiaso committed:
Fix flip/flop HttpRetryTest

testPostWithUnsupportedRetryAndFailureWhileRetrievingResponse

Testing retry scenario, the first POST request gets interrupted (e.g.
network cable loose connection during response), the second request gets
through. The test itself waited for the response of the second request
and checked if both requests were received by the remote server (also
locally set-up in this test). It seems like often the remote-server
also works through the requests in order (and adds something to
m_serverPostLog); however there are also multiple threads handling the
incoming requests, the remote server may as well handle the requests
out of order (e.g. first handle the second request, afterwards handle
the first one). But as we only wait for the second request/response the
servlet may not even have worked through the first request yet.
Solution: Wait until both requests have been received at the servlet
(timeout with 60 seconds in total if something went terribly wrong).

bschwarzent committed:
Make scroll containers unfocusable

Browser manufacturers recently changed the default behavior of
scrollable elements (which they call "scrollers"). To improve
accessibility, these elements are now automatically keyboard focusable
(but not click focusable) when the content is larger than the viewport
and does not contain any focusable elements.
See: https://developer.chrome.com/blog/keyboard-focusable-scrollers

This introduces additional tab stops that are currently not properly
visualized by Scout and may therefore be confusing. It also interferes
with existing logic in some of the more complex widgets, e.g. Table,
Tree, TileGrid. For example, the '.table' element has an explicit
tabindex="0" attribute, but the inner '.table-data' element does not.
However, because the '.table-data' is scrollable, the browser now makes
the element focusable automatically. But using the arrow keys to scroll
the '.table-data' will not work, because the Table widget uses those
keystrokes to move the select row and automatically refocuses the
'.table' element.

To restore the previous behavior, we automatically add a negative
tabindex to all scroll containers. This skips the element during key
navigation, but it may still receive the focus programmatically or via
a mouse click. Therefore, we use "-2" instead of "-1" (same native
behavior) and explicitly exclude it in the custom ':focusable' selector.
The FocusContext can therefore detect when such an element receives the
focus and can re-position the focus to the nearest focusable parent
element.

In future versions, we plan to make scrollable containers focusable
again, but it requires additional changes (e.g. proper handling of
keystrokes).

381281

bschwarzent committed:
objects: add sortProperties()

Add a method to copy of an object structure with the properties sorted
alphabetically (instead of insertion order). This can be useful when
formatting an object as JSON.

mvilliger committed:
Tree: use right and left cursor keystrokes to expand/collapse

- Tree/Outline: Use ← and → cursor for the same functionality as the
                existing plus and minus (both keystrokes are supported
                now and have the same functionality)
- Table: Fix displayText of ← and → cursor keystrokes in hierarchical
         tables.
- TreeCollapseOrDrillUpKeyStroke: Only collapse if child nodes are
                                  available. Otherwise, directly select
                                  the parent node. This prevents that
                                  one keystroke seems as it has no
                                  functionality.
- CompactTree: reveal selection after keystroke navigation
- CompactTree: fix left keystroke to not toggle between first two child
               nodes.

Required by ARIA tree view spec:
https://www.w3.org/WAI/ARIA/apg/patterns/treeview/

287717

matthiaso committed:
AbstractHttpTransportManager ensure initialization is run only once

395516

mvilliger committed:
Support for Java 21 runtime

368738

matthiaso committed:
Log output (debug/trace) for access control/permissions

387274

cguglielmo committed:
AccessControl.ts: let app handle bootstrap errors

With Microsoft Edge, if a tab is closed and reopened after the http
session has expired, the index.html is loaded from cache without
revalidating which starts the bootstrap requests. These requests
will fail and may return 401 (Unauthorized). In that case, the app would
trigger a page reload because the login site needs to be displayed.
Unfortunately, the status code gets lost for the permission call, so the
message "Permissions were not synchronized successfully." is shown and
the user has to manually reload the page.

395126

mvilliger committed:
UI Callback Support

Adds API to call functions on the UI (browser) and wait for the result.

389633

cguglielmo committed:
Session.ts: don't log a warning on session timeout for poll requests

A session timeout is nothing unusual, just log an info message.

395836

cguglielmo committed:
Add dataObjects.ts with functions to work with contributions

cguglielmo committed:
TableField: improve validation result and make it extensible

394805

cguglielmo committed:
objects: add isPojo()

The existing isPlainObject actually has the wrong name, it does not
check for *plain* objects, just for objects.
Adjusting the behavior would likely break existing code, so there is now
a new function called isPojo (short for plain old JavaScript object)
with the correct implementation.
The old function isPlainObject is now deprecated and delegates to a new
one called isObject with the same implementation.

Knicksaas committed:
Tooltip: resolve text key

When the tooltip text includes a nls text key, it was not resolved
because the resolveTextKeys was missing.

mvilliger committed:
Remove unnecessary import assignment

287717

fschinkel committed:
HybridManager: improve error handling

Exceptions are handled differently depending on their type (e.g.
VetoExceptions are displayed to the user). Therefore, do not create a
new exception with the error message but log an info and rethrow the
original exception.

cguglielmo committed:
TableField.ts: fix empty behavior, FormField.ts: provide _computeEmpty()

Empty is now a regular property, therefore a property change event is
triggered and _renderEmpty called whenever the property changes.

Migration: if a widget implements _updateEmpty, it should now implement
_computeEmpty instead.

The table field now implements _computeEmpty which only returns
true if there are table rows. This fixes the behavior of mandatory
table fields.

bschwarzent committed:
SmartField: add method to update lookupRow without changing value

If the lookup call changes dynamically (e.g. language switch, or
singular/plural) the currently displayed lookup row in a SmartField
might need to be updated. This can be achieved by setting the value to
null and the back to the previous value. However, this triggers
property change events for the 'value' property, which might have
unwanted effects. To only update the cached 'lookupRow' from the
current value without changing the value itself, a new method
updateLookupRow() is provided. It re-resolves the lookup row and
updates the display text accordingly.

369273

cguglielmo committed:
tileUtil: moveOtherTilesDown can lead to overlaps

A tile that spans columns must consider the bottom grid data of every
spanned column, not just the first one.

395112

matthiaso committed:
getGrantedPermissionLevel: Ignore null attribute

In 66924c59a7a8f899ee8ba51acf6f239300c19a1d a warning was introduced if
getGrantedPermissionLevel was called with a null input; however this
null input is valid after all, see javadoc on IPermissionCollection;
remove incorrect warning.

397579

bschwarzent committed:
Tile: do not trigger onLoadDataCancel() when tile is reloaded

When a tile is reloaded while the previous load call is still running,
the previous future is cancelled explicitly. However, because we
immediately schedule a new load future, it is not necessary to call the
onLoadDataCancel() callback. If that callback is used to show an error
message, it would even be wrong! To achieve this behavior, we set a
special execution hint on the previous future to indicate that the
cancellation happend explicitly and does not need to be handled by the
tile.

392970

mvilliger committed:
Improve XXE Check

The scan should find patterns where the code is separated to multiple
lines.

cguglielmo committed:
App: improve bootstrap error handling

The browser may cache the index.html but start the bootstrap requests
which may fail due to a session timeout (see previous commit).

A session timeout may be handled in several ways:
- The server returns 401 (Unauthorized) which typically is the case
  for /api calls.
- The server returns the login page  .
- The server returns 200 (Success) with an error object.
  This is used by Scouts json-based UI-protocol. Also, a UiAuthFilter
  may return such a response if the request expects a json response
  because a login page, which is html, could not be handled.

Also, a request can be executed in multiple ways, using ajax.ts or
$.ajax resulting in different error objects.

The bootstrap error handling must react appropriately to these cases.
The following cases were not handled until now:
- Calls to /api returning 401 (e.g. /api/permissions)
- Calls to /res/config-properties.json, /defaultValues returning 200
  with an error

395126

bschwarzent committed:
FocusContext: prevent scrolling when validating the focused element

When a DOM element receives the focus, the FocusContext checks if the
element is focusable according to the ':focusable' selector. If not,
the nearest focusable parent element is focused instead. This check is
necessary because not all elements that can be focused natively are
considered to be focusable by scout (special case: tabindex="-2",
see jquery-scout-selectors.js).

When changing the focus, the new element should be focused without
changing the scroll position. Otherwise, the visible elements might
jump unexpectedly. An element receiving the focus is expected to
already be at the correct position.

Additionally, the changing of the focus should not be executed
immediately in the 'focusin' event handler. Otherwise, other event
listeners (such as drag events) might not be fired. To prevent
flickering, the focus call is scheduled using requestAnimationFrame(),
which executes it right before the next repaint but after other
javascript code has been executed.

381281

bschwarzent committed:
FocusContext: only change focus for elements with tabindex="-2"

Certains elements are sometimes focusable, sometimes not. For example,
an <iframe> is only (keyboard) focusable if it does not contain a
focusable elements, e.g. only text. In that case we want to accept the
focus, because the text could not be selected by mouse otherwise.

The special "focus later" case is only explicitly needed for elements
that are focusable, but we never want the focus to be on. We use the
special tabindex "-2" to mark them.

381281

cguglielmo committed:
FocusManager: fix wrong focus after dragging an element

The global mouse down listener of the focus manager is responsible to
allow or prevent focus gain on mouse down. If an element is draggable,
the focus gain cannot be prevented without preventing the dragstart
event, too.

Therefore, the focus manager allows the focus gain but has to restore
the focus afterward, so that the previously focused element will be
focused again. This functionality broke with commit
f318d652e5b6ec5b028f1f08fe762ead278e3951 (#381281).

To fix it, the logic from FocusContext to prevent the focus gain on
elements with tabindex=-2 has to be integrated into the global
mouse down handler. One advantage of this approach is that such
elements now don't get the focus at all, not even temporary, except
if a draggable element is clicked. As explained, the focus gain must
be allowed and may temporarily be set on an element with tabindex=-2.

Also:
- removed unnecessary HTMLElement generic of the JQuery type because
  it is the default value
- simplified checks in isFocusableByMouse and isDraggable because
  closest also considers the current element not only the parents.

397963

cguglielmo committed:
Table: startCellEdit must not fail if row is not rendered or hiding

Use case:
- Filter is active which hides all rows
- table.focusCell() is called

Currently, an exception occurs because startCellEdit calls
table.$cell() which fails because $row is null.

397551

If startCell is called while the row is still hiding because of the
filtering, the exception does not occur immediately.
Instead, the popup is positioned incorrectly and an exception
occurs as soon as the popup is being repositioned (e.g. when the table
is layouted or scrolled).

The rendered check at CellEditorPopup.position is not required to
solve this error. It has been added to make it consistent with the super
class. And to make it more failsafe because startCellEdit may return
a popup that is not rendered.

350207

matthiaso committed:
Use XmlUtility for XML parsing

397720

bschwarzent committed:
Add Java representation for Switch widget (ToggleSwitch)

Note: classes and packages use the name "toggle switch", because
"switch" is a reserved keyword in Java.

398685

bschwarzent committed:
Cleanup code

cguglielmo committed:
TableField: don't return error status if a cell value is missing

The message of the error status should be cleared if there are multiple
invalid cells with different messages. Unfortunately, an error status
is created if mandatory cells have no value. In that case, only the
flag validByMandatory must be set to false.

394805

gabrielstalder committed:
Allow named deferred queue flush jobs for simpler debugging

- Adds parameter to name a deferred queue's flush job
- If no name's provided, the deferred queue's mandatory transaction
member ID is used for the flush job's name

bschwarzent committed:
Fix SmartFieldSpec

Adjust to new field event 'empty' introduced with commit
c96bcc357450663d972a8eb0c8b56d1ac901c597

mvilliger committed:
OWASP dependency check

cguglielmo committed:
UiNotifications: don't poll during tests

If a component subscribes for a ui notification topic, the ui
notification polling starts. These poll requests will fail and new poll
requests will be scheduled, which leads to a lot of noise.

During tests, ui notifications are now disabled by default. If a
specific test needs it to be enabled, the mock can be removed by
calling UiNotificationMock.unregister(). To not influence other specs,
it should be registered again afterward.

Also:
- uiNotifications is now an instance of the class UiNotifications,
  so it can be replaced using the object factory.
- UiNotificationSystem can now be replaced using the object factory.

388703

fschinkel committed:
Charts: increase opacity for points in line charts

The hover points in a line chart were hardly recognizable as there
opacity was very low. Increase the opacity to 60 in order to make them
more visible. It is not increased even further so that there is still a
recognizable difference between unchecked and checked points.

397045

mvilliger committed:
Update Vulnerable Third-Party Components

397727

Knicksaas committed:
Tree: Do not send unnecessary events to server

- When Tree is not checkable, do not send any nodesChecked events
- Do not send empty nodesChecked events at all

384727

bschwarzent committed:
UiNotifications: minor improvements

388703

ralph-steiner committed:
AbstractMapTransactionMember#commitChanges must detect commit conflicts

Before this change, if cache did not contain a value and two
transactions do at the same time a data change (calls invalidate on
cache), a commit conflict may have been missed because oldValue in
removeMap is null.

399399

mfi123 committed:
NextStepUid is not reset and blocks the process

On the FreeStep, the process is continued with a tile and an automatic
step follows in the background, which throws an error. The user lands on
the FreeStep again, but the nextStepUid is now already set so that the
process is blocked.

312244

ralph-steiner committed:
BasicCache#invalidate must always mark inserts dirty on invalidation

If multiple transactions did commit changes on same key at the same time
(AbstractMapTransactionMember#commitChanges concurrently processed)
cache inserts may have been committed because transaction were not
marked as dirty.

399399

fschinkel committed:
Improve DeferredOperationQueue

The queue processes elements immediately if there are more than its
batch size. If there are fewer elements it waits for a specified time
for more elements to come and processes a last batch after this delay.
Then, if the queue is empty, the flush job completes and is started
again after new elements are added to the queue.

In order to minimize the number of loops that are executed, check at the
beginning of each loop if the next batch and therefore the whole queue
is empty and complete the flush job if it is. Otherwise, the flush job
will wait the specified delay unnecessarily before completing.

The job itself used the run context created by the specified supplier.
This results in one run context for all batches that are executed. As a
queue may run forever, this can lead to an enormous amount of
executions that are not committed in the same transaction. In addition,
note that the job itself does not really need a run context as it just
schedules the batch executions, which are the elements here that need an
own run context.
Therefore, use an empty run context for the flush job and the run
context created by the supplier for the batch executions.

391689

fschinkel committed:
Table: fix _columnAtX()

Consider the insets of the table data when calculating the column offset
as otherwise the returned column may be wrong e.g. the table-data has a
padding.

fschinkel committed:
PlainTextEncoder: improve encode

The font-icon-regex matched only spans with one or more css classes set
but no other attributes. Improve the regex in order to match spans that
have e.g. an aria-label.

matthiaso committed:
IdFactoryTest: Invalid name for non-static of method

matthiaso committed:
IdFactory: Fix fallback branch for findOfByTypesMethod

IRootId does not have any generic parameter, however AbstractRootId does; use this class for fallback.

gabrielstalder committed:
Add missing abstract key word for AbstractTableCustomizerWrapper

gabrielstalder committed:
Make createRow rowModel parameter optional

The parameter rowModel is implemented as optional but not marked as
such. This leads to unnecessary transpiler warnings. This changes marks
the parameter as actually optional.

bschwarzent committed:
Add missing abstract keyword to abstract class

Sonar issues

Commit:
Update Apache Derby to 10.17.1.0

Part of the yearly dependency upgrade.

391416

bschwarzent committed:
Ignore bean with non-empty constructor

Prevents warning during startup: "Skipping bean candidate because it has
no empty or injectable constructor()."

csehlke committed:
Improve StreamUtility concat

- StreamUtility#concat can now handle streams of different type similar to Stream#concat
- StreamUtility#concat is now null safe

gabrielstalder committed:
Add styles utility to extract CSS colors fromm CSS attributes

bschwarzent committed:
FocusManager: fix JS error when collapsing draggable table rows

If a row in a hierarchical table is expanded or collapsed, the view port
is re-rendered, which means all rows are removed and rendered again. The
mousedown event still bubbles to the FocusManager with the original
element. That element is no longer part of the DOM and thus does not
have a parent element. In case of draggable tables, focusUtils
.restoreFocusLater() is called, which could not handle this case
and threw an error.

401646
